### PR TITLE
p4rt performance test application

### PR DESCRIPTION
**Pull request by satish153**
p4rt_perf_test is a test application designed to assess the performance of a P4Runtime server across various P4 profiles. The application's primary focus is on measuring the time required to program a specified number of entries.

_$IPDK_RECIPE/install/bin/p4rt_perf_test -t 1 -o 1 -n 4000000 -p 1
Total num of entries: 4000000
Number of threads: 1
Operation: 1
Test Profile: 1
Thread data - Core: 20 start_index: 0 num_entries: 4000000
count: 4000000
Num of entries added: 4000000
Time taken: 37.4511 seconds
number of entires per second: 106806_

**Note:**
p4rt_perf.cc/h and p4rt_perf_simple_l2_demo.cc/h are new coded. The code in the rest of the files is repurposed from ovs-p4rt module.
Programming with multiple threads doesn't really work as Stratum currently supports 'write' by only the master.

### CODE REVIEW

**Comment by n-sandeep** [comment]
Hi @satish153 , any sample command on how to test this performance will be helpful
There are these options "t:o:n:p:", looks like all testing scale is based on these options.

**Comment by satish153** [comment]
> Hi @satish153 , any sample command on how to test this performance will be helpful There are these options "t : o : n : p :", looks like all testing scale is based on these options.

Updated the description.

File: clients/p4rt-ctl-cpp/p4rt_perf_session.cc [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+  ::grpc::Status status =
+      session->Stub().Write(&context, write_request, &response);
+
+  return GrpcStatusToAbslStatus(status);
+}
+
+::p4::v1::TableEntry* SetupTableEntryToInsert(P4rtSession* session,
+                                              ::p4::v1::WriteRequest* req) {
+  req->set_device_id(session->DeviceId());
```
Comment (n-sandeep) [inline clients/p4rt-ctl-cpp/p4rt_perf_session.cc:159]: Only Update is of type repeated in this WriteRequest message. Before calling this API in loop, we can set device_id, election_id only once for any writerequest ?

snippet:

message WriteRequest {
  uint64 device_id = 1;
  uint64 role_id = 2 [deprecated=true];
  string role = 6;
  Uint128 election_id = 3;
  // The write batch, comprising a list of Update operations. The P4Runtime
  // server may arbitrarily reorder messages within a batch to maximize
  // performance.
  repeated Update updates = 4;

File: clients/p4rt-ctl-cpp/p4rt_perf_simple_l2_demo.cc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    mac_info.dst_mac[2] = (src_int >> 24) & 0xFF;
+    mac_info.dst_mac[3] = (src_int >> 16) & 0xFF;
+    mac_info.dst_mac[4] = (src_int >> 8) & 0xFF;
+    mac_info.dst_mac[5] = src_int & 0xFF;
+
+    PrepareSimpleL2DemoTableEntry(table_entry, mac_info, p4info,
+                                  t_data.oper == ADD);
+    count++;
+  }
+  auto sts = p4rt_perf::SendWriteRequest(session, write_request);
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf_simple_l2_demo.cc:78]: I gave some thought to this during implementation. 
I thought as inserting entries constitutes preparing the entries and sending them to the server the time taken to prepare the entry should also be considered.  Considering only the time for sending will give better numbers I'm ok with that as well.

File: clients/p4rt-ctl-cpp/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    p4rt_perf_session.cc
+    p4rt_perf_session.h
+    p4rt_perf_tls_credentials.cc
+    p4rt_perf_tls_credentials.h
+    p4rt_perf_simple_l2_demo.cc
+    p4rt_perf_util.cc
+)
+
+target_compile_options(p4rt_perf_test_o PRIVATE -O3)
+target_include_directories(p4rt_perf_test_o PRIVATE
```
Comment (ffoulkes) [inline clients/p4rt-ctl-cpp/CMakeLists.txt:29]: Why is this here? As a rule, you only need to specify a directory in the binary tree if you need to include a generated header file.

**Review by ffoulkes — COMMENTED** [review]
I'm going to go ahead and submit this initial review now. I still need to review the code itself.

File: clients/p4rt-ctl-cpp/p4rt_perf_simple_l2_demo.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf_simple_l2_demo.cc]: Done.

File: clients/CMakeLists.txt [hunk 4]
```diff
@@ -7,3 +7,4 @@
 add_subdirectory(gnmi-ctl)
 add_subdirectory(p4rt-ctl)
 add_subdirectory(sgnmi_cli)
+add_subdirectory(p4rt-ctl-cpp)
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/CMakeLists.txt]: Done

File: clients/p4rt-ctl-cpp/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+set(PROTO_INCLUDES
+    # Protobuf C++ header files.
+    ${PB_OUT_DIR}
+)
+
+#################
+# p4rt_perf_test #
+#################
+
+add_library(p4rt_perf_test_o OBJECT
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/CMakeLists.txt]: Done.

File: clients/p4rt-ctl-cpp/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    p4runtime_proto
+)
+
+
+add_executable(p4rt_perf_test
+    $<TARGET_OBJECTS:p4rt_perf_test_o>
+)
+
+target_link_libraries(p4rt_perf_test PUBLIC
+    absl::strings
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/CMakeLists.txt]: Done

File: clients/p4rt-ctl-cpp/p4rt_perf.cc [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+#include "absl/memory/memory.h"
+#include "p4rt_perf_session.h"
+#include "p4rt_perf_simple_l2_demo.h"
+#include "p4rt_perf_tls_credentials.h"
+
+#define MAX_THREADS 8
+#define CORE_BASE 20
+
+// globals
+TestParams test_params = {};
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf.cc]: Removed namespace as I felt it is overkill for a simple test application

File: clients/p4rt-ctl-cpp/p4rt_perf.cc [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+           core_id[index], thread_data[index].start,
+           thread_data[index].num_entries);
+  }
+}
+
+void RunPerfTest(int tid) {
+  using namespace p4rt_perf;
+
+  // Start a new client session.
+  auto status_or_session = P4rtSession::Create(absl::GetFlag(FLAGS_grpc_addr),
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf.cc]: using 'session' below to extract session from status_or_session

File: clients/p4rt-ctl-cpp/p4rt_perf.cc
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf.cc:1]: Done

File: clients/p4rt-ctl-cpp/CMakeLists.txt (hunk 3 above)
Comment (satish153) [reply clients/p4rt-ctl-cpp/CMakeLists.txt]: cleaned up CMakeList.txt.

File: clients/p4rt-ctl-cpp/p4rt_perf.h
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf.h:1]: Done.

File: clients/p4rt-ctl-cpp/p4rt_perf_session.cc (hunk 1 above)
Comment (satish153) [reply clients/p4rt-ctl-cpp/p4rt_perf_session.cc]: Done.

File: clients/p4rt_perf_test/p4rt_perf_test.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef P4RT_PERF_H
+#define P4RT_PERF_H
+
+#include <stdint.h>
+
+enum OPER { ADD = 1, DEL = 2, ADD_DEL = 3 };
```
Comment (ffoulkes) [inline clients/p4rt_perf_test/p4rt_perf_test.h:9]: Rule: No dead code

You never reference ADD_DEL. You should either implement it or delete it.

File: clients/p4rt_perf_test/p4rt_perf_test.h [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+struct ThreadInfo {
+  uint32_t tid;
+  uint32_t core_id;
+  uint64_t start;
+  uint64_t num_entries;
+  uint32_t oper;
+  double time_taken;
+};
+
+struct TestParams {
```
Comment (ffoulkes) [inline clients/p4rt_perf_test/p4rt_perf_test.h:22]: Huh. I didn't know you could initialize structure members in a header file and include the file in more than one compilation unit.

**Comment by ffoulkes** [comment]
Nuts. I meant to leave a clarifying comment when I closed my review.

The issues I'm chiefly concerned about (the "blockers") are the latent bugs:

- Missing input validation, especially where out-of-bounds values could lead to array overrun.
- Missing error returns, especially where continuing execution can lead to catastrophic failure.

Plus the compiler warnings.

File: clients/p4rt_perf_test/CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    absl::flags_private_handle_accessor
+    absl::flags
+    absl::statusor
+    absl::strings
+    p4runtime_proto
+    stratum_static
+    stratum_proto
+)
+
+install(TARGETS p4rt_perf_test DESTINATION bin)
```
Comment (ffoulkes) [inline clients/p4rt_perf_test/CMakeLists.txt:48]: We do normally build test programs (much less install them). We might want to set the EXCLUDE_FROM_ALL property on the `p4rt_per_test` target. The customer can build it on demand by issuing the command
```bash
cmake --build build --target p4rt_perf_test
```

File: clients/p4rt_perf_test/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+#
+
+include(CheckLibraryExists)
+
+set(PROTO_INCLUDES
+    # Protobuf C++ header files.
+    ${PB_OUT_DIR}
+)
+
+set(SRC_FILES
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/CMakeLists.txt:14]: You mean to say it is recommended to list the source files in add_executable instead of the CMake variable, SRC_FILES?

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+    core_id[index] = CORE_BASE + index;
+    thread_data[index].core_id = core_id[index];
+  }
+
+  // add remaining entries to the last thread;
+  thread_data[test_params.num_threads - 1].num_entries += rem_rules;
+
+  // dump the data
+  for (index = 0; index < test_params.num_threads; index++) {
+    printf("Thread data - Core: %d start_index: %d num_entries: %d\n",
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:82]: Done.

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+    case SIMPLE_L2_DEMO:
+      SimpleL2DemoTest(session.get(), p4info, t_data);
+      break;
+    default:
+      std::cerr << "unsupported profile " << std::endl;
+      return;
+  }
+}
+
+inline void PrintUsage(char* name) {
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:114]: Done.

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+  std::thread client_threads[MAX_THREADS];
+  for (int index = 0; index < test_params.num_threads; index++) {
+    client_threads[index] = std::thread(RunPerfTest, index);
+
+    /* Assign Thread Affinity */
+    CPU_ZERO(&cpuset);
+    CPU_SET(core_id[index], &cpuset);
+    if ((pthread_setaffinity_np(client_threads[index].native_handle(),
+                                sizeof(cpuset), &cpuset))) {
+      std::cerr << "error: pthread_setaffinity_np, rc:" << std::endl;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:175]: Done

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+      std::cerr << "unsupported profile " << std::endl;
+      return;
+  }
+}
+
+inline void PrintUsage(char* name) {
+  std::cerr << "Usage: " << name << " -t <value> -o <value> -n <value>"
+            << std::endl;
+  std::cout << "t: num of threads(optional, default: 1)" << std::endl;
+  std::cout << "o: operation(ADD=1, DEL=2)(mandatory)" << std::endl;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:118]: Done.

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+ * P4rt_perf_test is a test application designed to assess the performance
+ * of a P4Runtime server across various P4 profiles. The application's primary
+ * focus is on measuring the time required to program a specified number of
+ * entries.
+ *
+ * Command Line Arguments:
+ *   -t : Number of threads (optional, default: 1).
+ *   -o : Operation type (mandatory). 1 is to ADD and 2 to DEL
+ *   -n : Number of entries (optional, default: 1,000,000).
+ *   -p : P4 profile (optional, default: SIMPLE_L2_DEMO).
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:16]: Added a bit more description to make this clear.

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+        break;
+      case 'p':
+        test_params.profile = std::atoi(optarg);
+        break;
+      default:
+        PrintUsage(argv[0]);
+        return 1;
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:147]: made all input parameters unsigned and added the ValidateInput function to validate all the inputs.

File: clients/p4rt_perf_test/p4rt_perf_test.h (hunk 9 above)
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_test.h:9]: Done

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+           thread_data[index].num_entries);
+  }
+}
+
+void RunPerfTest(int tid) {
+  // Start a new client session.
+  auto status_or_session = P4rtSession::Create(absl::GetFlag(FLAGS_grpc_addr),
+                                               GenerateClientCredentials(),
+                                               absl::GetFlag(FLAGS_device_id));
+  if (!status_or_session.ok()) {
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:93]: Added error message. 
Can't return the status as this is a thread function. So updated the status in thread_data structure which is then checked in the main thread and returned as the status to the user. While this can be done better using 'future' and 'promise' features, I would think this is good enough for a standalone test application.

File: clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc [hunk 20]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#include "p4rt_perf_simple_l2_demo.h"
+
+#include "p4rt_perf_util.h"
+extern TestParams test_params;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc:7]: Done

File: clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+  uint64_t count = t_data.start + 1;
+
+  write_request.set_device_id(session->DeviceId());
+  *write_request.mutable_election_id() = session->ElectionId();
+  for (uint64_t j = 0; j < t_data.num_entries; j++) {
+    if (t_data.oper == ADD) {
+      table_entry = SetupTableEntryToInsert(session, &write_request);
+    } else if (t_data.oper == DEL) {
+      table_entry = SetupTableEntryToDelete(session, &write_request);
+    } else {
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc:58]: Implemented ValidateInput function which validates the inputs and returns on error. So no checks are needed here any more.

File: clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.h [hunk 22]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2023-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef P4RT_PERF_SIMPLE_L2_DEMO_H
+#define P4RT_PERF_SIMPLE_L2_DEMO_H
+
+#include "absl/flags/flag.h"
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.h:7]: Cleaned up the includes.

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    }
+  }
+
+  // Wait for all threads to finish
+  for (int index = 0; index < test_params.num_threads; index++) {
+    client_threads[index].join();
+  }
+
+  // evaluate and print perf numbers
+  // incase of multiple threads, use the maximum time taken by a thread to
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:186]: Done

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+    default:
+      std::cerr << "unsupported profile " << std::endl;
+      return;
+  }
+}
+
+inline void PrintUsage(char* name) {
+  std::cerr << "Usage: " << name << " -t <value> -o <value> -n <value>"
+            << std::endl;
+  std::cout << "t: num of threads(optional, default: 1)" << std::endl;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:117]: Done

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+  std::cout << "o: operation(ADD=1, DEL=2)(mandatory)" << std::endl;
+  std::cout << "n: num of entries(optional, default:1000000)" << std::endl;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:119]: Done

File: clients/p4rt_perf_test/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+    p4rt_perf_util.cc
+    p4rt_perf_util.h
+)
+
+add_executable(p4rt_perf_test ${SRC_FILES})
+
+target_compile_options(p4rt_perf_test PRIVATE -O3)
+
+target_include_directories(p4rt_perf_test PRIVATE ${PROTO_INCLUDES})
+
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/CMakeLists.txt:32]: Done

File: clients/p4rt_perf_test/CMakeLists.txt [hunk 27]
```diff
@@ -0,0 +1,48 @@
+# CMake build file for p4rt_perf_test
+#
+# Copyright 2023-2024 Intel Corporation
+# SPDX-License-Identifier: Apache 2.0
+#
+
+include(CheckLibraryExists)
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/CMakeLists.txt:7]: Removed

File: clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+  SimpleL2DemoMacInfo mac_info;
+  p4::v1::WriteRequest write_request;
+
+  int batch_size = t_data.num_entries;
+  uint64_t count = t_data.start + 1;
+
+  write_request.set_device_id(session->DeviceId());
+  *write_request.mutable_election_id() = session->ElectionId();
+  for (uint64_t j = 0; j < t_data.num_entries; j++) {
+    if (t_data.oper == ADD) {
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_simple_l2_demo.cc:54]: Done

File: clients/p4rt_perf_test/p4rt_perf_main.cc [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+                                               absl::GetFlag(FLAGS_device_id));
+  if (!status_or_session.ok()) {
+    return;
+  }
+
+  // Unwrap the session from the StatusOr object.
+  std::unique_ptr<P4rtSession> session = std::move(status_or_session).value();
+  ::p4::config::v1::P4Info p4info;
+  ::absl::Status status = GetForwardingPipelineConfig(session.get(), &p4info);
+  if (!status.ok()) return;
```
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/p4rt_perf_main.cc:101]: Added error message and updated status in thread_data.

File: clients/p4rt_perf_test/CMakeLists.txt (hunk 11 above)
Comment (satish153) [reply clients/p4rt_perf_test/CMakeLists.txt:48]: Leaving it to build and install with the standard build. We can revisit this if needed to exclude it.

File: clients/CMakeLists.txt
[the remark this answers is not in the document]
Comment (satish153) [reply clients/CMakeLists.txt:1]: Added documentation under docs/clients

File: clients/p4rt_perf_test/CMakeLists.txt (hunk 12 above)
[the remark this answers is not in the document]
Comment (satish153) [reply clients/p4rt_perf_test/CMakeLists.txt]: Done

**Review by ffoulkes — APPROVED** [review]
LGTM

The user guide needs some work, but that can be taken care of in a later commit.

**Comment by ffoulkes** [comment]
BTW, you need to do `git commit -s` to sign your commits.

The failed DCO check is a result of this.

**Comment by ffoulkes** [comment]
Change of plans (and Now for Something Completely Different): I've submitted a PR against your PR to bring the user guide in line with current documentation conventions.

**Review by n-sandeep — APPROVED** [review]
LGTM
